File: model-client/src/main/java/org/modelix/model/api/INode.java
package org.modelix.model.api;

import org.jetbrains.annotations.Nullable;

import java.util.stream.Stream;

public interface INode {
  boolean isValid();
  INodeReference getReference();

  IConcept getConcept();
  String getRoleInParent();
  INode getParent();

  Stream<INode> getChildren(String role);
  Stream<INode> getAllChildren();
  void addChild(String role, int index, INode node);
  INode addNewChild(String role, int index, @Nullable IConcept concept);
  void removeChild(INode child);

  INode getReferenceTarget(String role);
  void setReferenceTarget(String role, INode target);

  String getPropertyValue(String role);
  void setPropertyValue(String role, String value);
}

File: model-client/src/main/kotlin/org/modelix/model/api/INode.kt
package org.modelix.model.api

import java.util.stream.Stream

interface INode {
    val isValid: Boolean
    val reference: INodeReference?
    val concept: IConcept?
    val roleInParent: String?
    val parent: INode?
    fun getChildren(role: String?): Stream<INode?>?
    val allChildren: Stream<INode>
    fun addChild(role: String?, index: Int, node: INode?)
    fun addNewChild(role: String?, index: Int, concept: IConcept?): INode?
    fun removeChild(child: INode?)
    fun getReferenceTarget(role: String?): INode?
    fun setReferenceTarget(role: String?, target: INode?)
    fun getPropertyValue(role: String?): String?
    fun setPropertyValue(role: String?, value: String?)
}